blades/monopodws_imxrt.h

```diff
@@ -0,0 +1,143 @@
+#ifndef BLADES_MONOPODWS_H
+#define BLADES_MONOPODWS_H
+
+#ifdef ENABLE_WS2811
+
+#include <OctoWS2811.h>
+
+// Any group of digital pins may be used
+// Teensy 4.x can use any pin, or arbitrary group of pins!
+// default options {2, 14, 7, 8, 6, 20, 21, 5};
+const int numPins = 3;               //should be auto generated from config file
```

Comment:
This is also something that I might need a bit more help then just a suggestion :-)
When trying to alter the numPins and pinList during runtime, there were errors, indicating that a const can not be changed, and when making it a non const, the OctoWS2811 lib started complaining it expects a cont for numpins and pinlist.

So my hope was, to have some magic checking the config file pre-compile time and setting the numpins and pinlist maybe even via defines??  

Is it even possible during runtime that the frequency changes? I thought that all WS2811s use 800kHz?

What doe you mean by feeding the blades out in batches? I do not fully understand how that is supposed to work?

Reply:
Basically, I was thinking that we could do something like:

  bool CanUseCurrentBatch() {
     if (!engine.done()) return false;
     if (engine.hasPin(pin_)) return false;  // It already has this pin, feed out the data and start a new batch
     if (frequency_ != engine.frequency()) return false;
     // Check other parameters that must match here.
     return true;
  }

  bool IsReadyForEndFrame() {
      if (!engine.done()) return false;
      if (engine.num_pins() == 0) return true;
      if (CanUseCurrentBatch()) return true;
      // We've determined that we can't use the same batch, feed it out and wait until it's done.
      engine.show();
      return false;
  }

  void EndFrame() {
      while (!IsReadyForEndFrame());
      if (engine.num_pins() == 0) {
          engine.init(frequency_, and other paramters...)
      }
      int p = engine.add_pin(pin_, num_leds_);
      for (int i = 0; i < num_leds_; i++) { engine_.set_data(i, p, color_buffer[i]); }
      if (eigine.num_pins() == numPins) {
        // batch is full, feed out data now
        batch.show();
      }
  }

Basically, batches would be built and send out dynamically, with a maximum of "numPins" strings per batch.
This of course requires making some changes to OctoWS2811 so that pins can be added dynamically and init() can be called over and over again.


Now, pehrhaps a simpler option would be to completely forgoe the multi-string capabilities and do what monopodws does. Since the IMXRT code already have the capability of assembling the DMA data in as it goes, it should be simple to modify it to read directly from the color_buffer, just like the smt32 code does.  Of course, the reason the stm32 code does that is to save memory, which is not really relevant for this processor....

Reply:
Ah! Thank you for the explanation, now I understand it better. Only the numPins would still be a constant ste in the (hardware) config file? It would be nice if it was automatically set by the number Neopixel blades in the config file. Just using the NUM_BLADES define would not work, because there might be simple LED or Tri-cree blades in there as well.

Do you have an suggestion of how to derive the number of Neopixel blades from the config file, or during runtime?

Rewriting the OctoWS2811 to make it an "engine" that dynamically updates the pins, is something I can give a go. I really liked the idea of being able to leave the OctoWS2811 untouched. Seems that we need an modified version to be fully ProffieOS compliant.
I also understand better why you are re-initing the blade-pins and all parameters over and over. It makes it more flexible and will update the params if you change blades and have a drifferent blade(-ID) detected?

Reply:
> Do you have an suggestion of how to derive the number of Neopixel blades from the config file, or during runtime?

During compilation is difficult, or maybe impossible.
During runtime is easy, it's basically just doing a +=1 in the pin class constructor and a -=1 in the pin class destructor.